MCO on-cluster layering reorg the using OCL module (#93880)
openshift/openshift-docs: modules/coreos-layering-configuring-on-proc.adoc

// Module included in the following assemblies:
//
// * machine_configuration/mco-coreos-layering.adoc

:_mod-docs-content-type: PROCEDURE
[id="coreos-layering-configuring-on-proc_{context}"]
= Using on-cluster layering to apply a custom layered image

To apply a custom layered image to your cluster by using the on-cluster build process, create a `MachineOSConfig` custom resource (CR) that specifies the following parameters: 

* the Containerfile to build
* the machine config pool to associate the build
* where the final image should be pushed and pulled from
* the push and pull secrets to use

.Prerequisites

* You have the pull secret in the `openshift-machine-config-operator` namespace that the MCO needs in order to pull the base operating system image from your repository. By default, the MCO uses the cluster global pull secret, which it synchronizes into the `openshift-machine-config-operator` namespace. You can add your pull secret to the {product-title} global pull secret or you can use a different pull secret. For information on modifying the global pull secret, see "Updating the global cluster pull secret".

> * You have the push secret of the registry that the MCO needs to push the new custom layered image to. The credentials provided by the secret must also grant permission to delete an image from the registry.

Author: Resolved

.. Create the `MachineOSConfig` object:
+
[source,terminal]
----
$ oc create -f <file_name>.yaml
----

. If necessary, when the `MachineOSBuild` object has been created and is in the `READY` state, modify the node spec for the nodes where you want to use the new custom layered image:
+
.. Check that the `MachineOSBuild` object is `READY`. When the `SUCCEEDED` value is `True`, the build is complete.
+
[source,terminal]
----
$ oc get machineosbuild
----
+
.Example output showing that the `MachineOSBuild` object is ready
[source,terminal]
----
NAME                                                     PREPARED   BUILDING   SUCCEEDED   INTERRUPTED   FAILED   AGE
layered-image-ad5a3cad36303c363cf458ab0524e7c0-builder   False      False      True        False         False    43s
----

.. Edit the nodes where you want to deploy the custom layered image by adding a label for the machine config pool you specified in the `MachineOSConfig` object:
+
[source,terminal]
----
$ oc label node <node_name> 'node-role.kubernetes.io/<mcp_name>='
----
+
--
where:

node-role.kubernetes.io/<mcp_name>=:: Specifies a node selector that identifies the nodes to deploy the custom layered image. 
--
+

> When you save the changes, the MCO drains, cordons, and reboots the nodes. After the reboot, the node will be using the new custom layered image.

Reviewer: ```suggestion
When you save the changes, the MCO drains, cordons, and reboots the nodes. After the reboot, the node uses the new custom layered image.
```
Avoid use of future tense

layered-image-ad5a3cad36303c363cf458ab0524e7c0   False      True       False       False         False    12m <1> 
----
<1> The `MachineOSBuild` is named in the `<MachineOSConfig_CR_name>-<hash>` format.

. Verify that the `MachineOSConfig` object contains a reference to the new custom layered image by running the following command:
+
[source,terminal]
----
$ oc describe machineosconfig <object_name>
----
+
.Example digested image pull spec
[source,terminal]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineOSConfig
metadata:
  annotations:
    machineconfiguration.openshift.io/current-machine-os-build: layered-9a8f89455246fa0c42ecee6ff1fa1a45
  labels:
    machineconfiguration.openshift.io/createdByOnClusterBuildsHelper: ""
  name: layered-image
# ...
status:
  currentImagePullSpec: image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/os-image@sha256:3c8fc667adcb432ce0c83581f16086afec08a961dd28fed69bb6bad6db0a0754 <1>
----
<1> Digested image pull spec for the new custom layered image.

. Verify that the appropriate nodes are using the new custom layered image:

> .. Start a debug session as root for a control plane node:

Reviewer: ```suggestion
.. Start a debug session as root for a control plane node by running the following command:
```
